tests/mock_network/churn.rs

```diff
@@ -419,6 +410,70 @@ impl RelocationMapBuilder {
     }
 }
 
+// When do we send messages.
+#[derive(PartialEq, Eq)]
+enum MessageSchedule {
+    AfterChurn,
+    DuringChurn,
+}
+
+fn progress_and_verify<R: Rng>(
+    rng: &mut R,
+    network: &Network,
+    nodes: &mut [TestNode],
+    message_schedule: MessageSchedule,
+    added_indices: BTreeSet<usize>,
+    dropped_names: BTreeSet<XorName>,
+) {
+    if !dropped_names.is_empty() {
+        warn!("Dropping {:?}", dropped_names);
+    }
+
+    let (expectations, relocation_map) = match message_schedule {
+        MessageSchedule::AfterChurn => {
+            poll_and_resend(nodes);
+            let added_names = check_added_indices(nodes, added_indices);
+            log_churn_outcome(&added_names, &dropped_names);
+
+            let expectations = setup_expectations(rng, nodes, network.elder_size());
+            poll_and_resend(nodes);
+
+            (expectations, BTreeMap::default())
+        }
+        MessageSchedule::DuringChurn => {
+            let expectations = setup_expectations(rng, nodes, network.elder_size());
+            let relocation_map = RelocationMapBuilder::new(&nodes);
+
+            poll_and_resend(nodes);
+            let added_names = check_added_indices(nodes, added_indices);
+            log_churn_outcome(&added_names, &dropped_names);
+
+            (expectations, relocation_map.build(&nodes))
```

Comment:
would be nice to avoid the duplication, maybe worth putting that in a sub function `complete_churn` or better named.

```
             poll_and_resend(nodes);
            let added_names = check_added_indices(nodes, added_indices);
            log_churn_outcome(&added_names, &dropped_names); 
```

Reply:
That's already done in later commits.